lib/trie/database.go

```diff
@@ -185,22 +185,28 @@ func (t *Trie) loadNode(db Database, n *Node) error {
 	return nil
 }
 
-// PopulateMerkleValues writes the Merkle value of each children of the node given
-// as keys to the map merkleValues.
-func (t *Trie) PopulateMerkleValues(n *Node, merkleValues map[string]struct{}) {
-	if n.Kind() != node.Branch {
+// PopulateMerkleValues writes the Merkle values of the node given and of
+// all its descendant nodes as keys to the map merkleValues.
+// It is assumed the node and its descendant nodes have their Merkle value already
+// computed.
+func PopulateMerkleValues(n *Node, merkleValues map[string]struct{}) {
+	if n == nil {
 		return
 	}
 
-	branch := n
-	for _, child := range branch.Children {
-		if child == nil {
-			continue
-		}
+	if len(n.MerkleValue) == 0 {
```

Comment:
is n.MerkleValue always created/checked for validity before this call? Basically want to make sure that user input cannot cause this function to panic

Reply:
See https://github.com/ChainSafe/gossamer/pull/2871/files#r992189720

User input for now is only the offline pruner which guarantees all nodes have their cached Merkle value set.